Source/Core/Core/PowerPC/MMU.cpp

```diff
@@ -806,116 +940,143 @@ static __forceinline u32 TranslatePageAddress(const u32 address, const XCheckTLB
+				// a strange mapping. Need to check on hardware.
+				u32 address = (batl.BRPN | j) << 17;
+				// (BEPI | j) == (BEPI & ~BL) | (j & BL).
+				bat_table[batu.BEPI | j] = address | 0x1;
 			}
 		}
 	}
-	return false;
 }
 
-// Block Address Translation
-static u32 TranslateBlockAddress(const u32 address, const XCheckTLBFlag flag)
+static void UpdateFakeMMUDBat(u32 start_addr)
 {
-	u32 result = 0;
-	UReg_MSR& m_MSR = ((UReg_MSR&)PowerPC::ppcState.msr);
-	u32 batu = (m_MSR.PR ? BATU_Vp : BATU_Vs);
-
-	// Check for enhanced mode (secondary BAT enable) using 8 BATs
-	bool enhanced_bats = SConfig::GetInstance().m_LocalCoreStartupParameter.bWii && HID4.SBE;
-
-	if (flag != FLAG_OPCODE)
+	for (unsigned i = 0; i < (0x10000000 >> 17); ++i)
 	{
-		if (!CheckAddrBats(address, &result, batu, SPR_DBAT0U) && enhanced_bats)
-			CheckAddrBats(address, &result, batu, SPR_DBAT4U);
+		// Map from 0x4XXXXXXX or 0x7XXXXXXX to the range
+		// [0x7E000000,0x80000000).
+		u32 e_address = i + (start_addr >> 17);
+		u32 p_address = 0x7E000001 | ((i << 17) & Memory::FAKEVMEM_MASK);
+		dbat_table[e_address] = p_address;
 	}
-	else
+}
+
+void DBATUpdated()
+{
+	memset(dbat_table, 0, sizeof(dbat_table));
+	UpdateBATs(dbat_table, SPR_DBAT0U);
+	bool extended_bats = SConfig::GetInstance().m_LocalCoreStartupParameter.bWii && HID4.SBE;
+	if (extended_bats)
+		UpdateBATs(dbat_table, SPR_DBAT4U);
+	if (Memory::bFakeVMEM)
 	{
-		if (!CheckAddrBats(address, &result, batu, SPR_IBAT0U) && enhanced_bats)
-			CheckAddrBats(address, &result, batu, SPR_IBAT4U);
+		// In Fake-MMU mode, insert some extra entries into the BAT tables.
+		UpdateFakeMMUDBat(0x40000000);
+		UpdateFakeMMUDBat(0x70000000);
 	}
-	return result;
+	Memory::UpdateLogicalMemory(dbat_table);
+	JitInterface::ClearSafe();
+}
+
+void IBATUpdated()
+{
+	memset(ibat_table, 0, sizeof(ibat_table));
+	UpdateBATs(ibat_table, SPR_IBAT0U);
+	bool extended_bats = SConfig::GetInstance().m_LocalCoreStartupParameter.bWii && HID4.SBE;
+	if (extended_bats)
+		UpdateBATs(ibat_table, SPR_IBAT4U);
+	JitInterface::ClearSafe();
 }
 
 // Translate effective address using BAT or PAT.  Returns 0 if the address cannot be translated.
 template <const XCheckTLBFlag flag>
-u32 TranslateAddress(const u32 address)
+TranslateAddressResult TranslateAddress(const u32 address)
 {
-	// Check MSR[IR] bit before translating instruction addresses.  Rogue Leader clears IR and DR??
-	//if ((_Flag == FLAG_OPCODE) && !(MSR & (1 << (31 - 26)))) return _Address;
-
-	// Check MSR[DR] bit before translating data addresses
-	//if (((_Flag == FLAG_READ) || (_Flag == FLAG_WRITE)) && !(MSR & (1 << (31 - 27)))) return _Address;
-
-	// Technically we should do this, but almost no games, even heavy MMU ones, use any custom BATs whatsoever,
-	// so only do it where it's really needed.
-	if (SConfig::GetInstance().m_LocalCoreStartupParameter.bBAT)
+	u32 bat_result = (flag == FLAG_OPCODE ? ibat_table : dbat_table)[address >> 17];
```

Comment:
would doing bat_table[flag==FLAG_OPCODE][address >> 17] be any nicer? just a suggestion